### PR TITLE
Feature: URL Custom Attributes

### PR DESCRIPTION
**Pull request by udidol**
## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Description
An explanation of what is done in this PR

* Added a new Custom Attributes feature to the URL Control, which allows adding custom HTML attributes to a URL's `<a>` tag.
* Deprecated the URL control's `show_external` parameter.
* Replaced the `show_external` parameter with an `options` parameter (a flat array).
* Replaced all uses of `show_external` with an `options` array.

## Quality assurance

- [X] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: includes/base/element-base.php [hunk 1]
```diff
@@ -461,6 +461,17 @@ public function add_link_attributes( $element, array $url_control, $overwrite =
 			$attributes['rel'] = 'nofollow';
 		}
 
+		if ( ! empty( $url_control['custom_attributes'] ) ) {
+			// Custom URL attributes should come as a string of comma-delimited key|value pairs
+			$custom_attributes = explode( ',', $url_control['custom_attributes'] );
```
Comment (matipojo) [inline includes/base/element-base.php]: It should work like element custom attributes, each attribute in a separate line.

File: includes/controls/url.php [hunk 2]
```diff
@@ -64,13 +65,15 @@ public function get_default_value() {
...
-			'show_external' => true,
 			'placeholder' => __( 'Paste URL or type', 'elementor' ),
 			'autocomplete' => true,
+			'options' => [ 'is_external', 'nofollow', 'custom_attributes' ],
 			'dynamic' => [
 				'categories' => [ TagsModule::URL_CATEGORY ],
 				'property' => 'url',
 			],
+			'custom_attributes_description' => __( 'Set custom attributes for the anchor element. Separate attribute keys from values 
+			using the | (pipe) character. Separate key-value pairs with a comma. Do not use spaces.', 'elementor' ),
```
Comment (matipojo) [inline includes/controls/url.php]:
```suggestion
			using the | (pipe) character. Each attribute in a separate line. Do not use spaces.', 'elementor' ),
```

File: includes/controls/url.php [hunk 3]
```diff
@@ -86,14 +89,12 @@ protected function get_default_settings() {
...
-
 		$more_input_control_uid = $this->get_control_uid( 'more-input' );
-
 		$is_external_control_uid = $this->get_control_uid( 'is_external' );
-
 		$nofollow_control_uid = $this->get_control_uid( 'nofollow' );
+		$custom_attributes_uid = $this->get_control_uid( 'custom_attributes' );
 		?>
-		<div class="elementor-control-field elementor-control-url-external-{{{ data.show_external ? 'show' : 'hide' }}}">
+		<div class="elementor-control-field elementor-control-url-external-{{{ data.options ? 'show' : 'hide' }}}">
```
Comment (matipojo) [inline includes/controls/url.php]: Add backwards compatibility for controls that use `show_external`.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/base/element-base.php (hunk 1 above)
Comment (kobizz) [reply includes/base/element-base.php]: According to the plan, it should be in a single line

File: includes/controls/url.php [hunk 4]
```diff
@@ -64,13 +65,15 @@ public function get_default_value() {
 	protected function get_default_settings() {
 		return [
 			'label_block' => true,
-			'show_external' => true,
 			'placeholder' => __( 'Paste URL or type', 'elementor' ),
 			'autocomplete' => true,
+			'options' => [ 'is_external', 'nofollow', 'custom_attributes' ],
```
Comment (kobizz) [inline includes/controls/url.php]:
```suggestion
			'options' => [ 'external', 'nofollow', 'custom_attributes' ],
```

File: includes/controls/url.php [hunk 5]
```diff
@@ -113,6 +114,15 @@ public function content_template() {
 						<input id="<?php echo $nofollow_control_uid; ?>" type="checkbox" class="elementor-control-url-option-input" data-setting="nofollow">
 						<label for="<?php echo $nofollow_control_uid; ?>"><?php echo __( 'Add nofollow', 'elementor' ); ?></label>
 					</div>
+					<div class="elementor-control-url-option elementor-control-url__custom-attributes check-wrapper">
+						<label for="<?php echo $custom_attributes_uid; ?>" style="width: 65%;"><?php echo __( 'Custom Attributes', 'elementor' ); ?></label>
```
Comment (kobizz) [inline includes/controls/url.php]: Each style rule should be managed by an external CSS

**Review by kobizz — CHANGES_REQUESTED** [review]

File: includes/controls/url.php (hunk 2 above)
Comment (kobizz) [inline includes/controls/url.php]: Use trimming instead of forcing the user not to insert spaces

**Review by kobizz — CHANGES_REQUESTED** [review]

File: includes/controls/url.php (hunk 2 above)
Comment (udidol) [reply includes/controls/url.php]: trimming is ok, but people can put spaces in their keys/values. trimming doesn't solve that problem. I will add trimming, but I do think that part of the description should stay

**Pull request closed without merging**